smolyay/smolyak.py

```diff
@@ -0,0 +1,141 @@
+            raise IndexError(
+                "index_per_level must be an array with a"
+                "length of at least {}"
+                .format(exactness + 1))
+
+    def level_indexes(self):
+        """Generate indexes of levels.
+
+        Returns
+        -------
+        level_indexes : list
+            Unique indexes at each level.
+
+        """
+        # creating an array to store unique indexes in each level
+        level_indexes = [[0]]*(self.exactness + 1)
+        max_index = sum(self.index_per_level[0:self.exactness + 2])
+        indexes = np.linspace(0, max_index, max_index+1, dtype=np.int32)
+
+        # intial level
+        level_indexes[0] = indexes[0:self.index_per_level[0]]
+
+        # storing unique indexes for each level
+        summ_index = self.index_per_level[0]
+        for idx_index_per_level in range(1, self.exactness + 1):
+            summ_index += self.index_per_level[idx_index_per_level]
+            level_indexes[idx_index_per_level] = (
+                indexes[summ_index-self.index_per_level[idx_index_per_level]:
+                        summ_index])
+
+        return level_indexes
+
+    def grid_point_index(self):
+        """Generate index of grid points.
+
+        First, it generates allowed compositons of levels
+        via NEXCOM algorithm. Then, it replaces the levels
+        with their corresponding indexes, and
+        makes the index for the grid points.
+
+        Returns
+        -------
+        grid_points_index : numpy array
+            Index of grid points
+
+        """
+        num_grid_points = 0
+        level_indexes = self.level_indexes()
+        level_compositions = []
+
+        # compute the allowed level compositions
+        for summ in range(self.dimension, self.exactness+self.dimension+1):
+
+            # NEXCOM
+            max_level_index = summ - self.dimension
+            num_output = scipy.special.comb(max_level_index+self.dimension-1,
+                                            self.dimension-1, exact=True)
+            compositions = np.zeros((num_output, self.dimension),
+                                    dtype=np.int32)
+
+            # (A) first entry
+            r = np.zeros(self.dimension, dtype=np.int32)
+            r[0] = max_level_index
+            t = max_level_index
+            h = 0
+            compositions[0] = r
+            index = 1
+
+            # (D): these termination conditions should be redundant
+            while (r[self.dimension-1] != max_level_index
+                   and index < num_output + 1):
+
+                # (B)
+                if t != 1:
+                    h = 0
+
+                # (C)
+                h += 1
+                t = r[h-1]
+                r[h-1] = 0
+                r[0] = t-1
+                r[h] += 1
+
+                # (D)
+                compositions[index] = r
+                index += 1
+
+            # precompute the number of grid points
```

Comment:
I have a really hard time following the code from here onward, but I think that what you are doing can be achieved in a simpler way using `numpy.meshgrid`. See https://stackoverflow.com/questions/1208118/using-numpy-to-build-an-array-of-all-combinations-of-two-arrays for an explanation:

You would define `points = None` at the start of the function, then do something like this inside the for loop:
```py
            # get all combinations of points at each level
            for composition in compositions:
                point_ids = [level_indexes[i] for i in composition]
                points_ = np.array(np.meshgrid(*point_ids)).T.reshape(-1, self.dimension)
                if points is None:
                    points = points_
                else:
                    points = np.concatenate((points,points_), axis=0)
```
`points` will be a NumPy array, so you can either convert to a list using the `tolist()` method or leave as is

Reply:
Done.